File: archive/legacy-python/backend/app/storage.py
from __future__ import annotations

import shutil
from dataclasses import dataclass
from pathlib import Path

from fastapi import UploadFile


@dataclass(frozen=True)
class JobPaths:
    root: Path
    uploads: Path
    inputs: Path
    outputs: Path
    terrain_root: Path
    terrain_mbtiles: Path
    tilejson: Path
    stylejson: Path
    base_mbtiles: Path


@dataclass(frozen=True)
class TilesetPaths:
    root: Path
    mbtiles: Path


class Storage:
    def __init__(self, root: Path) -> None:
        self.root = root
        self.jobs_root = root / "jobs"
        self.tilesets_root = root / "tilesets"
        self.jobs_root.mkdir(parents=True, exist_ok=True)
        self.tilesets_root.mkdir(parents=True, exist_ok=True)

    def paths_for(self, job_id: str) -> JobPaths:
        root = self.jobs_root / job_id
        uploads = root / "uploads"
        inputs = root / "inputs"
        outputs = root / "outputs"
        terrain_root = outputs / "terrain"
        paths = JobPaths(
            root=root,
            uploads=uploads,
            inputs=inputs,
            outputs=outputs,
            terrain_root=terrain_root,
            terrain_mbtiles=outputs / "terrain-rgb.mbtiles",
            tilejson=outputs / "tiles.json",
            stylejson=outputs / "style.json",
            base_mbtiles=uploads / "base.mbtiles",
        )
        for path in [root, uploads, inputs, outputs, terrain_root]:
            path.mkdir(parents=True, exist_ok=True)
        return paths

    def tileset_paths_for(self, tileset_id: str) -> TilesetPaths:
        root = self.tilesets_root / tileset_id
        root.mkdir(parents=True, exist_ok=True)
        return TilesetPaths(root=root, mbtiles=root / "tiles.mbtiles")

    async def save_upload(self, upload: UploadFile, destination: Path) -> Path:
        destination.parent.mkdir(parents=True, exist_ok=True)
        with destination.open("wb") as output:
            while chunk := await upload.read(1024 * 1024):
                output.write(chunk)
        await upload.close()
        return destination

    def clear_inputs(self, paths: JobPaths) -> None:
        if paths.inputs.exists():
            shutil.rmtree(paths.inputs)
        paths.inputs.mkdir(parents=True, exist_ok=True)

File: kotlin/terrain-web/src/main/kotlin/com/terrainconverter/web/Storage.kt
package com.terrainconverter.web

import io.ktor.http.content.PartData
import io.ktor.utils.io.core.readBytes
import java.nio.file.Files
import java.nio.file.Path
import java.nio.file.StandardCopyOption
import kotlin.io.path.createDirectories
import kotlin.io.path.deleteIfExists
import kotlin.io.path.exists

data class JobPaths(
    val root: Path,
    val uploads: Path,
    val inputs: Path,
    val outputs: Path,
    val terrainRoot: Path,
    val terrainMbtiles: Path,
    val tilejson: Path,
    val stylejson: Path,
    val baseMbtiles: Path,
)

data class TilesetPaths(
    val root: Path,
    val mbtiles: Path,
)

class Storage(val root: Path) {
    private val jobsRoot = root.resolve("jobs")
    private val tilesetsRoot = root.resolve("tilesets")

    init {
        jobsRoot.createDirectories()
        tilesetsRoot.createDirectories()
    }

    fun pathsFor(jobId: String): JobPaths {
        val root = jobsRoot.resolve(jobId)
        val uploads = root.resolve("uploads")
        val inputs = root.resolve("inputs")
        val outputs = root.resolve("outputs")
        val terrainRoot = outputs.resolve("terrain")
        listOf(root, uploads, inputs, outputs, terrainRoot).forEach { it.createDirectories() }
        return JobPaths(
            root = root,
            uploads = uploads,
            inputs = inputs,
            outputs = outputs,
            terrainRoot = terrainRoot,
            terrainMbtiles = outputs.resolve("terrain-rgb.mbtiles"),
            tilejson = outputs.resolve("tiles.json"),
            stylejson = outputs.resolve("style.json"),
            baseMbtiles = uploads.resolve("base.mbtiles"),
        )
    }

    fun tilesetPathsFor(tilesetId: String): TilesetPaths {
        val root = tilesetsRoot.resolve(tilesetId)
        root.createDirectories()
        return TilesetPaths(root = root, mbtiles = root.resolve("tiles.mbtiles"))
    }

    suspend fun saveUpload(part: PartData.FileItem, destination: Path) {
        destination.parent?.createDirectories()
        Files.write(destination, part.provider().readBytes())
        part.dispose()
    }

    fun moveFile(source: Path, destination: Path) {
        destination.parent?.createDirectories()
        Files.move(source, destination, StandardCopyOption.REPLACE_EXISTING)
    }

    fun clearInputs(paths: JobPaths) {
        if (paths.inputs.exists()) {
            Files.walk(paths.inputs)
                .sorted(Comparator.reverseOrder())
                .forEach { if (it != paths.inputs) it.deleteIfExists() }
        }
        paths.inputs.createDirectories()
    }
}
